### ISSUE TITLE
Distributed Store `get` doesn't work well with `add`

### ISSUE THREAD

**Issue by ejguan** [pytorch/pytorch#80349]
### 🐛 Describe the bug

While I was working on sharing distributed seed for DataLoader, internal users encountered a problem that the distributed process will hang when trying to `get` an integer counter from the distributed store (`PrefixStore`).

While I dived deeper on this problem, I found a comment here:
https://github.com/pytorch/pytorch/blob/590d3e5774110e4657dcaa6acdb387ef69e41b47/torch/distributed/distributed_c10d.py#L242-L246

Do we have a plan to remove the legacy store to make sure `get` is working with `add`? If not, should we add some notes to recommend users to use `add(key, 0)` to retrieve the value saved by `store.add`?
https://pytorch.org/docs/stable/distributed.html#torch.distributed.Store.add

### Versions

fblearner

**Comment by rohan-varma** [comment]
@ejguan Thanks for filing the issue! Do you have a reproduction (is it as simple as calling add() followed by get()?) of the issue, if so that would be great to share. 

cc @H-Huang for any thoughts.

**Comment by ejguan** [comment]
This is reported by internal users. See [link](https://fburl.com/jnvixeog). You can reproduce the Error if you back out the following [Diff](https://fburl.com/diff/s92k2g5r)

It seems only happening for users who are using fblearner with PrefixStore

**Comment by rohan-varma** [comment]
Bumping priority as it can affect internal workloads
